[SPARK-25003][PYSPARK] Use SessionExtensions in Pyspark

RussellSpitzer opened the pull request:
Master

## What changes were proposed in this pull request?

Previously Pyspark used the private constructor for SparkSession when
building that object. This resulted in a SparkSession without checking
the sql.extensions parameter for additional session extensions. To fix
this we instead use the Session.builder() path as SparkR uses, this
loads the extensions and allows their use in PySpark.

## How was this patch tested?

An integration test was added which mimics the Scala test for the same feature.

Please review http://spark.apache.org/contributing.html before opening a pull request.

RussellSpitzer committed:
[SPARK-25003][PYSPARK] Use SessionExtensions in Pyspark

Previously Pyspark used the private constructor for SparkSession when
building that object. This resulted in a SparkSession without checking
the sql.extensions parameter for additional session extensions. To fix
this we instead use the Session.builder() path as SparkR uses, this
loads the extensions and allows their use in PySpark.

SparkQA commented:
**[Test build #94149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94149/testReport)** for PR 21990 at commit [`f790ae0`](https://github.com/apache/spark/commit/f790ae000ae1c3f4030162028186448d345e2984).

SparkQA commented:
**[Test build #94149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94149/testReport)** for PR 21990 at commit [`f790ae0`](https://github.com/apache/spark/commit/f790ae000ae1c3f4030162028186448d345e2984).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94149/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #94151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94151/testReport)** for PR 21990 at commit [`84c2513`](https://github.com/apache/spark/commit/84c2513a2cb7a50ba01cdc0d49e2ad7f9583799b).

SparkQA commented:
**[Test build #94151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94151/testReport)** for PR 21990 at commit [`84c2513`](https://github.com/apache/spark/commit/84c2513a2cb7a50ba01cdc0d49e2ad7f9583799b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94151/
Test PASSed.

HyukjinKwon commented:
@RussellSpitzer, let's close other ones except for this and name it `[SPARK-25003][PYSPARK] Use SessionExtensions in Pyspark`. Let me review this one within few days. Also, I don't think we should do it, at least, to branch-2.2. This logic here is quite convoluted and I would rather avoid to backport even to branch-2.3 actually.

HyukjinKwon commented on python/pyspark/sql/session.py:
```diff
@@ -218,7 +218,9 @@ def __init__(self, sparkContext, jsparkSession=None):
-                jsparkSession = self._jvm.SparkSession(self._jsc.sc())
+                jsparkSession = self._jvm.SparkSession.builder() \
+                    .sparkContext(self._jsc.sc()) \
+                    .getOrCreate()
```
@RussellSpitzer, mind checking the logic `getOrCreate` inside Scala side and deduplicate them here while we are here? Some logics for instance setting default session, etc. are duplicated Here in Python side and there in Scala side.

It would be nicer if we have some tests as well. `spark.sql.extensions` are static configuration, right? in that case, we could add a test, for example, please refer https://github.com/apache/spark/pull/21007. I added a test with static configuration before there.

HyukjinKwon commented:
@RussellSpitzer, also please ping me here if you face any difficulties. am wiling to help or push some changes to your branch.

RussellSpitzer committed:
[SPARK-25003][PYSPARK]: Add Tests for spark.sql.extensions in Pyspark

Adds a test which sets spark.sql.extensions to a custom extension
class. This is the same as the SparkExtensionsSuite which does the
same thing in Scala.

SparkQA commented:
**[Test build #94866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94866/testReport)** for PR 21990 at commit [`7ba70b5`](https://github.com/apache/spark/commit/7ba70b524f9779529142f6c70b04610b5b068a05).

SparkQA commented:
**[Test build #94870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94870/testReport)** for PR 21990 at commit [`d5c37b7`](https://github.com/apache/spark/commit/d5c37b732f1948d8240bd8de33a080ac5db03571).

SparkQA commented:
**[Test build #94870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94870/testReport)** for PR 21990 at commit [`d5c37b7`](https://github.com/apache/spark/commit/d5c37b732f1948d8240bd8de33a080ac5db03571).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkExtensionsTest(unittest.TestCase, SQLTestUtils):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94870/
Test FAILed.

SparkQA commented:
**[Test build #94873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94873/testReport)** for PR 21990 at commit [`0eea205`](https://github.com/apache/spark/commit/0eea205ca0591c68975412873b34393f6bf19437).

SparkQA commented:
**[Test build #94873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94873/testReport)** for PR 21990 at commit [`0eea205`](https://github.com/apache/spark/commit/0eea205ca0591c68975412873b34393f6bf19437).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkExtensionsTest(unittest.TestCase, SQLTestUtils):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94873/
Test PASSed.

RussellSpitzer commented on python/pyspark/sql/tests.py:
```diff
@@ -3563,6 +3563,51 @@ def test_query_execution_listener_on_collect_with_arrow(self):
+            .appName(cls.__name__) \
+            .config(
+                "spark.sql.extensions",
+                "org.apache.spark.sql.MyExtensions") \
```
I'm not sure what you mean here? This class already exists as part of the SparkSqlExtensions Suite and i'm just reusing with no changes.

HyukjinKwon replied:
Oh sorry I thought you wrote that class for this test.

RussellSpitzer replied:
Reduce, reuse, recycle :)

RussellSpitzer commented on python/pyspark/sql/tests.py:
```diff
@@ -3563,6 +3563,51 @@ def test_query_execution_listener_on_collect_with_arrow(self):
+        cls.spark.stop()
+
+    def tearDown(self):
+        self.spark._jvm.OnSuccessCall.clear()
```
Sounds good to me, i'll take that out.

SparkQA commented:
**[Test build #94975 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94975/testReport)** for PR 21990 at commit [`21ff627`](https://github.com/apache/spark/commit/21ff6273a7fcf54c2f40abde951d4aab89f96a02).

SparkQA commented:
**[Test build #94975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94975/testReport)** for PR 21990 at commit [`21ff627`](https://github.com/apache/spark/commit/21ff6273a7fcf54c2f40abde951d4aab89f96a02).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkExtensionsTest(unittest.TestCase, SQLTestUtils):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94975/
Test PASSed.

RussellSpitzer commented:
@HyukjinKwon So i've been staring at this for a while today, and I guess the big issue is that we always need to make a Python SparkContext to get a handle on the JavaGateway, so everything that happens before the context is made cannot just be a wrapper of SQL methods and must reimplement. Unless we decided to refactor the code so that the JVM is more generally available (probably not possible) we will be stuck with redoing code in python ...

HyukjinKwon commented on python/pyspark/sql/session.py:
```diff
@@ -218,7 +218,9 @@ def __init__(self, sparkContext, jsparkSession=None):
                 jsparkSession = self._jvm.SparkSession.getDefaultSession().get()
             else:
-                jsparkSession = self._jvm.SparkSession(self._jsc.sc())
+                jsparkSession = self._jvm.SparkSession.builder() \
```
@RussellSpitzer, have you maybe had a chance to take a look and see if we can deduplicate some logics comparing to Scala's `getOrCreate`? I am suggesting this since now it looks the code path duplicates some logics there.

HyukjinKwon reviewed (commented):
@RussellSpitzer, does that mean it's difficult to deduplicate the logics between here and `getOrCreate` from Scala side? Will take a close look soon anyway.

This looks better be addressed since the current change actually executes quite duplicated code paths there ..

RussellSpitzer commented:
What I wanted was to just call the Scala Methods, instead of having half the code and half in python, but we create the JVM in the SparkContext creation code so this ends up not being a good method I think. We could just translate the rest of GetOrCreate into Python but then every time there is a patch of the code in scala it will need a Python mod as well.

HyukjinKwon commented:
> We could just translate the rest of GetOrCreate into Python but then every time there is a patch of the code in scala it will need a Python mod as well.

@RussellSpitzer, I actually think we already duplicate some codes in Python side and Scala side at this code path now (see `getOrCreate` at Scala side and `__init__` at Python side). They are now duplicatedly executed with this change. I believe this isn't a orthogonal stuff to handle separately.

If that requires a bit of duplicated codes in Python side to avoid duplicated code path executions, then I think that can be a workaround to get through for now.

SparkQA commented:
**[Test build #96199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96199/testReport)** for PR 21990 at commit [`def4f3e`](https://github.com/apache/spark/commit/def4f3e91b38e7d63c03fc7206489fd01220e878).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkExtensionsTest(unittest.TestCase, SQLTestUtils):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96199/
Test PASSed.

RussellSpitzer committed:
SPARK-25003: Add helper methods to create new Extensions from Conf

Previously the only way to add extensions to the session was via the
getOrCreate method of the SparkSession Builder. To facilitate non-scala
Session creation we add a new constructor which takes in just the
context and Extensions. Then we also add a new Extensions constructor
which given a SparkConf generates an Extensions object with user config
already applied.

SparkQA commented:
**[Test build #96257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96257/testReport)** for PR 21990 at commit [`67d9772`](https://github.com/apache/spark/commit/67d9772e1f470bba83e311cd10cfe586a7f11b92).

SparkQA commented:
**[Test build #96257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96257/testReport)** for PR 21990 at commit [`67d9772`](https://github.com/apache/spark/commit/67d9772e1f470bba83e311cd10cfe586a7f11b92).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96257/
Test PASSed.

holdenk commented:
I'm +1 on switching to the builder and not using the private interface.

RussellSpitzer commented:
I'm fine with anything really, I still think the ideal solution is probably not to tie the creation of the py4j gateway to the SparkContext, but that's probably a much bigger refactor.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala:
```diff
@@ -169,3 +178,29 @@ class SparkSessionExtensions {
+   * Initialize extensions if the user has defined a configurator class in their SparkConf.
+   * This class will be applied to the extensions passed into this function.
+   */
+  private[sql] def applyExtensionsFromConf(conf: SparkConf, extensions: SparkSessionExtensions) {
```
Can you just put this in `SparkSession` object and make it `private[spark]`?

HyukjinKwon commented on python/pyspark/sql/session.py:
```diff
@@ -212,13 +212,17 @@ def __init__(self, sparkContext, jsparkSession=None):
         self._sc = sparkContext
         self._jsc = self._sc._jsc
         self._jvm = self._sc._jvm
+
```
really not a big deal but let's revert this newline to leave related changes only.

HyukjinKwon commented:
Looks close to go.

HyukjinKwon commented on python/pyspark/sql/session.py:
```diff
@@ -212,13 +212,17 @@ def __init__(self, sparkContext, jsparkSession=None):
                 jsparkSession = self._jvm.SparkSession.getDefaultSession().get()
             else:
-                jsparkSession = self._jvm.SparkSession(self._jsc.sc())
+                extensions = self._sc._jvm.org.apache.spark.sql\
```
tiny nit: just for consistency `extensions = self._sc._jvm.org.apache.spark.sql\` -> `extensions = self._sc._jvm.org.apache.spark.sql \`

RussellSpitzer committed:
SPARK-25003: Address Reviewer Comments

SparkQA commented:
**[Test build #96824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96824/testReport)** for PR 21990 at commit [`8a26728`](https://github.com/apache/spark/commit/8a2672806c6513692435bef113935f2e6f67a910).

SparkQA commented:
**[Test build #96824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96824/testReport)** for PR 21990 at commit [`8a26728`](https://github.com/apache/spark/commit/8a2672806c6513692435bef113935f2e6f67a910).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96824/
Test FAILed.

SparkQA commented:
**[Test build #96825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96825/testReport)** for PR 21990 at commit [`27c9d28`](https://github.com/apache/spark/commit/27c9d28430d5c0cc5404947d9ec690d53a626c2d).

SparkQA commented:
**[Test build #96825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96825/testReport)** for PR 21990 at commit [`27c9d28`](https://github.com/apache/spark/commit/27c9d28430d5c0cc5404947d9ec690d53a626c2d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96825/
Test FAILed.

SparkQA commented:
**[Test build #96827 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96827/testReport)** for PR 21990 at commit [`cf7cf75`](https://github.com/apache/spark/commit/cf7cf75efc616151d75e1da10630731fa98e117e).

SparkQA commented:
**[Test build #96828 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96828/testReport)** for PR 21990 at commit [`4ddaff8`](https://github.com/apache/spark/commit/4ddaff8fcd1bfedc221eda7d6f37e5e9f28bfb4e).

SparkQA commented:
**[Test build #96828 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96828/testReport)** for PR 21990 at commit [`4ddaff8`](https://github.com/apache/spark/commit/4ddaff8fcd1bfedc221eda7d6f37e5e9f28bfb4e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96828/
Test PASSed.

HyukjinKwon commented on python/pyspark/sql/session.py:
```diff
@@ -219,6 +219,7 @@ def __init__(self, sparkContext, jsparkSession=None):
                 jsparkSession = self._jvm.SparkSession.getDefaultSession().get()
             else:
                 jsparkSession = self._jvm.SparkSession(self._jsc.sc())
+
```
Oh haha, let's get rid of this change

RussellSpitzer replied:
I'm addicted to whitespace apparently

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -86,6 +86,7 @@ class SparkSession private(
 
   private[sql] def this(sc: SparkContext) {
```
@RussellSpitzer, mind if I ask to note that this is currently not used in Scala code base but PySpark and tests in Scala side?

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -1136,4 +1121,27 @@ object SparkSession extends Logging {
+   * Initialize extensions if the user has defined a configurator class in their SparkConf.
+   * This class will be applied to the extensions passed into this function.
+   */
+  private[sql] def applyExtensionsFromConf(conf: SparkConf, extensions: SparkSessionExtensions) {
```
Let's make it `private`

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -86,6 +86,7 @@ class SparkSession private(
 
   private[sql] def this(sc: SparkContext) {
     this(sc, None, None, new SparkSessionExtensions)
+    SparkSession.applyExtensionsFromConf(sc.getConf, this.extensions)
```
Let's add some comments why this is only here in this constructor. It might look weird why this constructor specifically requires to run `applyExtensionsFromConf` alone.

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -3563,6 +3563,48 @@ def test_query_execution_listener_on_collect_with_arrow(self):
                 "The callback from the query execution listener should be called after 'toPandas'")
 
 
+class SparkExtensionsTest(unittest.TestCase, SQLTestUtils):
```
I think `SQLTestUtils` is not needed.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -1136,4 +1121,27 @@ object SparkSession extends Logging {
+   * This class will be applied to the extensions passed into this function.
+   */
+  private[sql] def applyExtensionsFromConf(conf: SparkConf, extensions: SparkSessionExtensions) {
+    val extensionConfOption = conf.get(StaticSQLConf.SPARK_SESSION_EXTENSIONS)
```
I think we can even only pass `conf.get(StaticSQLConf.SPARK_SESSION_EXTENSIONS)` as its argument instead of `SparkConf`, and name it `applyExtensions`.

HyukjinKwon reviewed (approved):
LGTM otherwise

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -1136,4 +1121,27 @@ object SparkSession extends Logging {
+   * Initialize extensions if the user has defined a configurator class in their SparkConf.
+   * This class will be applied to the extensions passed into this function.
+   */
+  private[sql] def applyExtensionsFromConf(conf: SparkConf, extensions: SparkSessionExtensions) {
```
How about returning `SparkSessionExtensions` from this method, and modifying the secondary constructor of `SparkSession` as:

```scala
private[sql] def this(sc: SparkContext) {
  this(sc, None, None,
    SparkSession.applyExtensionsFromConf(sc.getConf, new SparkSessionExtensions))
}
```

I'm a little worried whether the order we apply extensions might affect.

ueshin replied:
On second thoughts, we could move the method call to the top of the default constructor?

RussellSpitzer replied:
The Default constructor of SparkSession?

RussellSpitzer replied:
I thought about this and was worried then about multiple invocations of the extensions
Once every time the SparkSession is cloned

ueshin replied:
Oh, I see, moving to the default constructor was not a good idea.
How about the first suggestion?

ueshin replied:
I see, but in that case, we need to ensure that no injection of extensions is used in the default constructor to avoid initializing without extensions from the conf.

HyukjinKwon replied:
Eh .. I think it's okay to have a function and returns that updated extensions.

HyukjinKwon replied:
Actually either way looks okay.

RussellSpitzer replied:
Updated with replacement then :)

RussellSpitzer committed:
SPARK-25003: More Refactoring

Removes SparkConf from applyExtensions, now only accepts a Optional
string which can contain a classname for extensions. Removed errant
whitespace.

SparkQA commented:
**[Test build #97471 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97471/testReport)** for PR 21990 at commit [`8ab76d4`](https://github.com/apache/spark/commit/8ab76d4385d1849c43ffd940d539c143e0ba8c81).

SparkQA commented:
**[Test build #97471 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97471/testReport)** for PR 21990 at commit [`8ab76d4`](https://github.com/apache/spark/commit/8ab76d4385d1849c43ffd940d539c143e0ba8c81).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97471/
Test FAILed.

RussellSpitzer commented:
Addressed Comments from @HyukjinKwon , I'm interested in @ueshin 's suggestions, but I can't figure out how to do that unless we bake it into the Extensions constructor. If we place it in the Sessions constructor then invocations of "newSession" will reapply existing extensions. I added a note in the code.

SparkQA commented:
**[Test build #97472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97472/testReport)** for PR 21990 at commit [`d9b2a55`](https://github.com/apache/spark/commit/d9b2a55275b74c406d9f9c435bf1b53a6ef4b35a).

SparkQA commented:
**[Test build #97472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97472/testReport)** for PR 21990 at commit [`d9b2a55`](https://github.com/apache/spark/commit/d9b2a55275b74c406d9f9c435bf1b53a6ef4b35a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97472/
Test PASSed.

RussellSpitzer committed:
SPARK-25003: Refactor ApplyExtensions Function

It now returns the extensions in modifies

SparkQA commented:
**[Test build #97497 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97497/testReport)** for PR 21990 at commit [`3629c78`](https://github.com/apache/spark/commit/3629c78118eb77589ae1126e877f0fd2b57441ce).

SparkQA commented:
**[Test build #97497 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97497/testReport)** for PR 21990 at commit [`3629c78`](https://github.com/apache/spark/commit/3629c78118eb77589ae1126e877f0fd2b57441ce).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97497/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #97510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97510/testReport)** for PR 21990 at commit [`3629c78`](https://github.com/apache/spark/commit/3629c78118eb77589ae1126e877f0fd2b57441ce).

ueshin commented:
LGTM.

SparkQA commented:
**[Test build #97510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97510/testReport)** for PR 21990 at commit [`3629c78`](https://github.com/apache/spark/commit/3629c78118eb77589ae1126e877f0fd2b57441ce).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97510/
Test PASSed.

HyukjinKwon commented:
Merged to master.

shiyuhang0 commented:
Why not port it to Spark < 3
